File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class TrieNode 
{
    TrieNode[] children;
    boolean isWord;
    
    public TrieNode()
    {
        this.children = new TrieNode[26];
        this.isWord = false;
    }
}

class WordDictionary {
    TrieNode root;
    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        TrieNode node = root;
        for(int i = 0; i < word.length(); i++)
        {
            char c = word.charAt(i);
            if(node.children[c - 'a'] == null)
            {
                node.children[c - 'a'] = new TrieNode();
            }
            node = node.children[c - 'a'];
        }
        
        node.isWord = true;
    }
    
    private boolean dfs(int j, TrieNode curt, String word)
    {
        TrieNode node = curt;
        for(int i = j; i < word.length(); i++)
        {
            char c = word.charAt(i);
            if(c == '.')
            {
                for(int k = 0; k < 26; k++)
                {
                    if(node.children[k] != null && dfs(i + 1, node.children[k], word)) return true;
                }
                return false;
            }
            else if(node.children[c - 'a'] == null) return false;
            
            node = node.children[c - 'a'];
        }
        
        return node.isWord;
    }
    
    public boolean search(String word) {
        TrieNode node = root;
        return dfs(0, node, word);
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.cpp
class TrieNode{
public:
    bool isWord;
    TrieNode* next[26];
    
    TrieNode()
    {
        this->isWord = false;
        for(int i = 0; i < 26; i++)
        {
            this->next[i] = nullptr;
        }
    }
};

class WordDictionary {
public:
    TrieNode* root = new TrieNode();
    WordDictionary() {
        
    }
    
    void addWord(string word) {
        TrieNode* cursor = root;
        
        for(auto& c : word)
        {
            if(cursor->next[c - 'a'] == nullptr)
            {
                cursor->next[c - 'a'] = new TrieNode();
            }
            
            cursor = cursor->next[c - 'a'];
        }
        
        cursor->isWord = true;
    }
    
    bool dfs(int i, TrieNode* cursor, string word)
    {
        if(i == word.length()) return cursor->isWord;
        
        if(word[i] == '.')
        {
            bool valid = false;
            for(int j = 0; j < 26; j++)
            {
                if(cursor->next[j] != nullptr)
                {
                    if(dfs(i + 1, cursor->next[j], word))
                    {
                        return true;
                    }
                }
            }
            return false;
        }
        else
        {
            if(cursor->next[word[i] - 'a'] != nullptr)
                return dfs(i + 1, cursor->next[word[i] - 'a'], word);
            else
            {
                return false;
            }
        }
    }
    
    bool search(string word) {
        TrieNode* cursor = root;
        return dfs(0, cursor, word);
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */